### PR TITLE
Add crypto module

### PR DESCRIPTION
**Pull request by parfeon**
feat(crypto): add crypto module

Add crypto module that allows to configure SDK to encrypt and decrypt messages.

fix(crypto): fix legacy cryptor

Improved security of crypto implementation by adding enhanced AES-CBC cryptor.

### CODE REVIEW

File: lib/pubnub/modules/crypto/crypto_module.rb [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        return nil if header.nil?
+
+        cryptor_identifier = header.identifier || '\x00\x00\x00\x00'
+        cryptor = cryptor cryptor_identifier
+
+        # Check whether there is a cryptor to decrypt data or not.
+        if cryptor.nil?
+          identifier = header.identifier || 'UNKN'
+          raise UnknownCryptorError, {
+            message: "Decrypting data created by unknown cryptor. Please make sure to register
```
[the remark this answers is not in the document]
Comment (parfeon) [reply lib/pubnub/modules/crypto/crypto_module.rb:100]: @seba-aln what typo?

File: lib/pubnub/modules/crypto/cryptor_header.rb [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+      # * sentinel
+      # * version
+      # * cryptor identifier
+      # * cryptor data size
+      # * cryptor-defined fields size.
+      def length
+        # Legacy payload doesn't have header.
+        return 0 if @data.nil?
+
+        9 + (data_size < 255 ? 1 : 3)
```
[the remark this answers is not in the document]
Comment (parfeon) [reply lib/pubnub/modules/crypto/cryptor_header.rb:147]: It returns the size of the header (how many bytes encode header information) – it used to compute offsets in data and pull out crypto-defined data followed by encrypted data.

File: lib/pubnub/formatter.rb [hunk 3]
```diff
@@ -42,16 +42,40 @@ def format_uuid(uuids, should_encode = true)
...
       # Transforms message to json and encode it
-      def format_message(message, cipher_key = "", use_random_iv = false, uri_escape = true)
-        if cipher_key && !cipher_key.empty?
-          pc = Pubnub::Crypto.new(cipher_key, use_random_iv)
-          message = pc.encrypt(message).to_json
-          message = Addressable::URI.escape(message) if uri_escape
-        else
-          message = message.to_json
-          message = Formatter.encode(message) if uri_escape
+      # def format_message(message, cipher_key = '', use_random_iv = false, uri_escape = true)
```
[the remark this answers is not in the document]
Comment (parfeon) [reply lib/pubnub/formatter.rb:45]: @seba-aln good catch! I forgot to clean it up.

**Review by seba-aln — APPROVED** [review]

**Comment by parfeon** [comment]
@pubnub-release-bot release

**Pull request merged**